### PR TITLE
sevice-loadbalancer: First take at a sv lb diagram

### PR DESCRIPTION
**Pull request by jayunit100**
Heres a first take at a diagram for describing how this works for users.
Also hoping it will help clarify details which i may be missing.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/contrib/136)

### CODE REVIEW

**Comment by jayunit100** [comment]
@bprashanth , does this look accurate to you ?

**Comment by bprashanth** [comment]
It's a little confusing, haproxy can directly proxy to service endpoints, it doesn't need services with nodeports though your diagram seems to indicate it does. I can make more sensible recommendations when I have some more time. 

In somewhat related news: 
1. Are you up for testing servicelb? 
2. We should: mv service-loadbalancer haproxy; mkdir service-loadbalancers; mv haproxy service-loadbalancers.

@aledbf might also have some thoughts

**Comment by jayunit100** [comment]
Yes, up for testing, happy to put those changes in that you mentioned also

**Comment by jayunit100** [comment]
@bprashanth look closer :) the serviceloadbalancer is on the RIGHT, the OLD way is on the left... sorry i shuda labelled em

now makes more sense :) ?

**Comment by bprashanth** [comment]
Oh, I missed that. Well I can't really comment on this, can I ask for an ascii rendition instead? that way we can go back and forth. Doesn't have to be fancy.

**Comment by jayunit100** [comment]
ok will do graphviz ascii

**Comment by jayunit100** [comment]
Ok @bprashanth , https://github.com/kubernetes/contrib/pull/136/files Updated with diagram generated from the ASCII text .

**Comment by aledbf** [comment]
@jayunit100 what are you trying to you in the graph?

**Comment by aledbf** [comment]
@jayunit100 I think the graph could be hard to follow.

I'm assuming the traffic reaches the same node with `role=loadbalancer`

`User -> service (VIP) -> kube-proxy -> go LB -> flannel -> pod/s`
with
`User -> haproxy VIP -> haproxy -> flannel -> pod/s`

Using haproxy to forward the traffic to the service removes functionality like controlling the algorithm to be used.

**Comment by jayunit100** [comment]
Yes, @aledbf  you are correct, but  
- ..As per the docs `HAProxy is the only working implementation` , right? , so this diagram shows how a concrete use of service loadbalancers works.  
- I've updated it to clarify that :) .... 

does this make a little more sense?

**Comment by jayunit100** [comment]
@aledbf ping.. does it make more sense now that i've updated the diagram to clarify the "loadbalancer" vs the old way of doing things?  

I think we are all on the same page, but the diagram may have been misleading originally.

**Comment by jayunit100** [comment]
Hi folks, I've updated the README and figure as well to clarify things.

File: service-loadbalancer/service-loadbalancer.dot [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        user->haproxyLB [label="Service Loadbalancer: Pluggable + pod-aware loadbalancing alternative\n which forwards directly to endpoints."];
+
+} 
+        subgraph cluster_a { 
+        label="Nodeports";
+         haproxy->k1nodeport_8001;
+         haproxy->k2nodeport_8001;
+         k1nodeport_8001->service;
+         k2nodeport_8001->service;
+         service->k1PostgreSQL [label=targetPort_80];
```
Comment (jayunit100) [inline service-loadbalancer/service-loadbalancer.dot]: nit.  possibly this could be **outside** the nodeport box?    just a thought I'd put in here as  part of a self-review im doing.

**Comment by jayunit100** [comment]
Hi @bprashanth since this is just a doc update, 
- Any thoughts on merging this as a first pass doc on how it works?  
- Or do you think its misleading to people?   
I think it could really help to jumpstart the conversation around it ...

**Comment by bprashanth** [comment]
I would still prefer raw ascii :)
I can cat the file, view it in terminal, copy/paste diagram into github comment etc. I'd only resort to graphing tools when the design is set in stone. Even the kubernetes architecture diagram is inaccurate, and no one has updated it. However I'll take anything with boxes and straight lines, for the sake of consistency with the rest of our docs.

I would also not shove both those before/after views of the world into the same diagram. Ideally something like:
Short explanation about what the first diagram is (without using specific terms like haproxy and postgres)
ascii
A short explanation about why the setup is sub optimal.
Segway into new setup with serviceloadbalacer
ascii 

And finally, I would number the steps like we do in services.md (https://github.com/kubernetes/kubernetes/blob/f00e0b4b18cbed8442c6b76134b9a6503cd29646/docs/user-guide/services.md#ips-and-vips)

**Comment by jayunit100** [comment]
Ok.. I see what you mean. So, how do you diagram things in ascii? If you give me an example I can take a look. I've never done an ascii diagram before... Or are you suggesting no diagram? I'm also happy to close the PR if you think it's not a good idea......

**Comment by jayunit100** [comment]
Ah ! I found http://asciiflow.com which I can try

**Comment by bprashanth** [comment]
Ascii flow looks useful. No real format for ascii, that's why it's useful, just do something like https://lwn.net/Articles/620087/ or what I did in the proposal https://github.com/kubernetes/kubernetes/pull/12827/files

**Comment by jayunit100** [comment]
Ok @bprashanth hows this ... 
- pure ascii . 
- removed the "before" vs "after" diagram style

File: service-loadbalancer/README.md [hunk 2]
```diff
@@ -15,6 +15,35 @@ This module will not help with the latter
 
 ## Overview
 
+The image below roughly describes the service-loadbalancer alternative (compared with using NodePorts).
```
Comment (bprashanth) [inline service-loadbalancer/README.md]: This sentence is confusing. Suggest removing and moving this diagram down to the section where we discuss tcp (since it's actually tcp specific).

File: service-loadbalancer/README.md [hunk 3]
```diff
@@ -15,6 +15,35 @@ This module will not help with the latter
...
+
+```
+                          +----------------------+                     
+                          | haproxy load balancer|
```
Comment (bprashanth) [inline service-loadbalancer/README.md]: this box is actually service loadbalancer, not haproxy. labeling it service loadbalancer implies some lb is being used, but the reverse is not true.

File: service-loadbalancer/README.md [hunk 4]
```diff
@@ -15,6 +15,35 @@ This module will not help with the latter
...
+
+```
+                          +----------------------+                     
+                          | haproxy load balancer|                     
+                          +---------+------------+                     
+                                    |                                  
+                +-------------------+---------------+                  
+                |                                   |                  
+   +------------+----------------+     +-----------+--------------------+
+   |kubelet role=loadbalancer (1)|     |kubelet role=loadbalancer (2)   |
```
Comment (bprashanth) [inline service-loadbalancer/README.md]: make this node role=loadbalancer since one cannot assign labels to the kubelet

File: service-loadbalancer/README.md [hunk 5]
```diff
@@ -15,6 +15,35 @@ This module will not help with the latter
...
+   +----+------------------------+     +--------------------------------+
+        |                                                       |
+        |                                                       |
+        | --tcpServices=postgresql:3333                         | --tcpServices=postgresql:3333
```
Comment (bprashanth) [inline service-loadbalancer/README.md]: you started the loadbalancer with --tcpServices right? so why not put this up there in the haproxy box?

File: service-loadbalancer/README.md [hunk 6]
```diff
@@ -15,6 +15,35 @@ This module will not help with the latter
...
+        |                                                       |
+        | --tcpServices=postgresql:3333                         | --tcpServices=postgresql:3333  
+        |                                                       |    
+        |               +-----------------------------+         |    
+        |               |   service for postgresql    |         |    
+        |               +-----------------------------|         |    
+        |                      |  targetPort=2222,port=3333     |    
+        |           +------+-----------+    +----------------+  |    
+        |           |                  |    |                |  |    
+        +---------->| k1postgresql     |    | k2postgresql   | <-
```
Comment (bprashanth) [inline service-loadbalancer/README.md]: Since the tcp example is actually mysql it would help if you reworded this example to suit the example

File: service-loadbalancer/README.md [hunk 7]
```diff
@@ -15,6 +15,35 @@ This module will not help with the latter
...
+        |               |   service for postgresql    |         |    
+        |               +-----------------------------|         |    
+        |                      |  targetPort=2222,port=3333     |    
+        |           +------+-----------+    +----------------+  |    
+        |           |                  |    |                |  |    
+        +---------->| k1postgresql     |    | k2postgresql   | <-  
+                    | container        |    | container      |
+                    | Port=2222        |    | Port=2222      |          
+                    +------------------+    +----------------+          
+Direct load balanced connections  to kubelets, without need for NodePort on every kubelet.
```
Comment (bprashanth) [inline service-loadbalancer/README.md]: The kubelet is nowhere in this picture actually, the loadbalancer opens up a tcp connection to the endpoint

File: service-loadbalancer/README.md (hunk 4 above)
Comment (jayunit100) [inline service-loadbalancer/README.md]: ah interesting i didn't realize the distinction

File: service-loadbalancer/README.md (hunk 7 above)
Comment (jayunit100) [inline service-loadbalancer/README.md]: This is a good point, by endpoint, you mean the containerPort, correct?  Since that container is running in a conceptual "pod", that is why its shown this way.  Maybe I'm missing something?

Comment (bprashanth) [inline service-loadbalancer/README.md]: I meant the container can keep serving traffic even if the kubelet is dead

**Comment by bprashanth** [comment]
Bump @jayunit100

**Comment by jayunit100** [comment]
Hiya beeps.will look at it this week and update the diagram for sure !  got pulled in a few other efforts

**Comment by jayunit100** [comment]
I've updated with the above comments.  Sorry about the delay ; have been pulled away on other stuff.
Am happy to close or handoff the patch to someone else if anyone wants to take over

**Comment by test-foxish** [comment]
recomputing cla status...

**Comment by test-foxish** [comment]
recomputing cla status...

**Comment by jayunit100** [comment]
closing
